File: week09/20/pr3_Maximum_Depth_of_N-ary_tree.py

# Definition for a Node.
class Node:
    def __init__(self, val: int = None, children: list['Node'] = None):
        self.val = val
        self.children = children

from collections import deque

class Solution:
    def maxDepth(self, root: 'Node') -> int:
        # Make a stack of nodes
        # Increase cnt + 1 whenever it calls children
        
        # No.. go recursive
        if not root : return 0
        elif not root.children : return 1
        lst = deque()
        def helper(node,cnt):
            cnt+=1
            if not node.children :
                lst.append(cnt)
                return
            else :
                n = len(node.children)
                for j in range(n):
                    helper(node.children[j],cnt)

        for i in range(len(root.children)):
            cnt = 1
            helper(root.children[i],cnt)
        return max(lst)
        

File: week09/20/pr3_Maximum_Depth_of_N-ary_tree.c

//  Definition for a Node.
struct Node {
      int val;
      int numChildren;
      struct Node** children;
 };
 

int max(int a, int b){
    return a >= b ? a : b;
}

int maxDepth(struct Node* root) {
    if(!root) return 0;

    int cnt = 0;
    for(int i=0;i<root->numChildren;i++){
        cnt = max(cnt,maxDepth(root->children[i]));
    }
    return cnt+1;
}